### ISSUE TITLE
[Metadata] Configurable AssertionConsumerService isDefault and index values (IndexedEndpointType)

### ISSUE THREAD

**Issue by peppelinux** [IdentityPython/pysaml2#599]
In my SP metadata I have
`<ns0:AssertionConsumerService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST" Location="http://sp1.test.org:8000/saml2/acs/" index="1"/>`

I'd like instead to have:
`<ns0:AssertionConsumerService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST" Location="http://sp1.test.org:8000/saml2/acs/" index="0" isDefault=True/>`

I tried a little modification in md.AssertionConsumerService (line 520) this way:
````
class AssertionConsumerService(IndexedEndpointType_):
    """The urn:oasis:names:tc:SAML:2.0:metadata:AssertionConsumerService
    element """
    c_tag = 'AssertionConsumerService'
    c_namespace = NAMESPACE
    c_children = IndexedEndpointType_.c_children.copy()
    c_attributes = IndexedEndpointType_.c_attributes.copy()
    c_child_order = IndexedEndpointType_.c_child_order[:]
    c_cardinality = IndexedEndpointType_.c_cardinality.copy()

    def __init__(self, **kwargs):
        super().__init__()
        self.index = 0
        self.is_default = "true"
````

But still cannot set the index value to zero.
More in the deep I also see that...

I've also found in metadata.do_attribute_consuming_service, at line 485, index value as a constant:
````
        ac_serv = md.AttributeConsumingService(
            index="1", service_name=service_name,
            requested_attribute=requested_attributes)
````
`metadata.do_attribute_consuming_service` also take `spsso` object as argument, it have `spsso.assertion_consumer_service[0]` with the followings attributes (see index):
````
{'extension_attributes': {}, 
 'location': 'http://sp1.testunical.it:8000/saml2/acs/', 
 'text': None, 
 'index': '1', 
 'binding': 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST',
 'encrypted_assertion': None,
 'extension_elements': [],
 'response_location': None,
 'is_default': None}
````
Probably `ac_serv` could use `spsso.assertion_consumer_service[0]`'s index value to get coupled with it, rather then a constant value like "1".

`spsso` originates in `metadata.do_spsso_descriptor` and it seems that the index value is initialized in `metadata.do_endopints`, as following:
````
def do_endpoints(conf, endpoints):
    service = {}

    for endpoint, (eclass, indexed) in endpoints.items():
        try:
            servs = []
            i = 1
````

Setting the initial value of `i` to 0 and changing the constant argument "index" in md.AttributeConsumingService to `0` made finally the index starting from 0...

Bytheway I'd prefer a solution with any modifications in the pySAML2 code.
I patched in my SP implementation the descriptor objects, before the metadata get encoded in the http response, in this way:
````
def metadata(request, config_loader_path=None, valid_for=None):
    """Returns an XML with the SAML 2.0 metadata for this
    SP as configured in the settings.py file.
    """
    conf = get_config(config_loader_path, request)
    metadata = entity_descriptor(conf)

    # this will renumber acs starting from 0 and set index=0 as is_default
    cnt = 0
    for attribute_consuming_service in metadata.spsso_descriptor.attribute_consuming_service:
        attribute_consuming_service.index = str(cnt)
        cnt += 1

    cnt = 0
    for assertion_consumer_service in metadata.spsso_descriptor.assertion_consumer_service:
        assertion_consumer_service.is_default = 'true' if not cnt else '' 
        assertion_consumer_service.index = str(cnt)
        cnt += 1
        
    return HttpResponse(content=text_type(metadata).encode('utf-8'),
                        content_type="text/xml; charset=utf8")
````

My question is, is there any way to do this using specific options in SP_CONF? 
Why ACS start from 1?
Is there an option to configure a ACS binding as the default one (isDefault)?

I read [OASIS Specifications, saml-metadata-2.0-os.pdf - "2.2.3 Complex Type IndexedEndpointType"](https://docs.oasis-open.org/security/saml/v2.0/saml-metadata-2.0-os.pdf) for better view and I understand that pySAML2 behaviour is formally correct. This question is due to the fact the italian SPID federation needs this kind of metadata to permit us to join! 

Thank you

**Comment by peppelinux** [comment]
I found a partial solution for assertion_consumer_service index, following Documentation [here](https://github.com/IdentityPython/pysaml2/blob/master/docs/howto/config.rst#endpoints).

````
    'service': {
        'sp': {
                'assertion_consumer_service': [
                    ('%s/acs/' % BASE_URL, saml2.BINDING_HTTP_POST, 0),
                    ],
````

This permit me to set the index value but there's need also to set which resource isDefault.
What's your ideas about the possibility to add this attribute as kwarg in sp_config's endpoint configuration file?

Regarding `attribute_consuming_service` I cannot get to set indexed value from endpoint configuration.
Any suggestions or ideas about this? Thank you

**Comment by peppelinux** [comment]
If I change the ACS idenx value as the previous example SATOSA get this error

"Incorrect request from requester: Unknown entity or unsupported bindings"

Watching the AuthnRequest I found

````
<ns0:AuthnRequest AssertionConsumerServiceURL="('http://sp1.testunical.it:8000/saml2/acs/', 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST', 1)"
````
`urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST', 1)` looks very strange!

**Comment by peppelinux** [comment]
I faced this problem coding a more low level solution, an example here:
https://github.com/peppelinux/Django-Identity/blob/36cf34a3d7f1e10edf7414c350f5025a599e5b8f/djangosaml2_sp/djangosaml2_sp/djangosaml2_spid/views.py#L290

Probably would be better if pysaml2 removes these contants value in the code and move to a better generalized (and configurable) approach, in this field.

**Comment by peppelinux** [comment]
The faulty documentation is here:
https://github.com/IdentityPython/pysaml2/blob/dc7be03999b90e9b9131ae1ffa84a8986c719d6b/docs/howto/config.rst#endpoints

at paragraph "or a 3-tuple (URL+binding+index)"

**Comment by peppelinux** [comment]
This issue shown as solved in pySAML2 v6.4.0
Great!

**Comment by peppelinux** [comment]
It seems fixed in metadata but not in authreq sso service.

Example, configuring a sp with:
````
            'endpoints': {
                'assertion_consumer_service': [
                    (f'{BASE_URL}/acs/', saml2.BINDING_HTTP_POST, 1),
                    ],
````

Auth IDP will fail, here the debugging info:
![pysaml_acs_url](https://user-images.githubusercontent.com/1297620/103152077-b30a8b80-4784-11eb-8997-121837dae82c.png)

Removing the idnex number, like:

````
            'endpoints': {
                'assertion_consumer_service': [
                    (f'{BASE_URL}/acs/', saml2.BINDING_HTTP_POST),
                    ],
````

it works
![2](https://user-images.githubusercontent.com/1297620/103152089-cb7aa600-4784-11eb-9555-cdb344d7ab01.png)
